### PR TITLE
fix: team flow (cf https://github.com/pathwar/pathwar/issues/618)

### LINKED ISSUE #618: Implement new team flow
Be careful, the `Season` model requires a new field named `SubscriptionStatus` with possible values: `Open`, `Close` (later we will probably have new values, i.e., `EditOnly` to support adding members but not creating a new team)

```
                                                            ╔════════════════════════════╗                                                                                                
                                                            ║                            ║                                                                                                
                                                            ║    can I join a season?    ║                                                                                                
                                                            ║                            ║                                                                                                
                                                            ╚════════════════════════════╝                                                                                                
                                                                           │                                                                                                              
                                                                           │                                                                                                              
                                                                           ▼                                                                                                              
                                                            ┌────────────────────────────┐          ┌───────────────────┐                                                                 
                                                            │                            │          │ Nope, you cannot  │                                                                 
                                                            │Is the season status "Open" │          │  join a private   │                                                                 
                                                            │       or "Unlisted"?       │── no ───▶│      season       │                                                                 
                                                            │                            │          │                   │                                                                 
                                                            └────────────────────────────┘          └───────────────────┘                                                                 
                                                                           │                                                                                                              
                                                                         yes                                                                                                              
                                                                           ▼                                                                                                              
                                                            ┌────────────────────────────┐           ┌──────────────────┐                                                                 
                                                            │                            │           │                  │                                                                 
                                                            │  I already have a team in  │           │ Nope, max 1 team │                                                                 
                                                            │        that season         │─── yes ──▶│    per season    │                                                                 
                                                            │                            │           │                  │                                                                 
                                                            └────────────────────────────┘           └──────────────────┘                                                                 
                                                                           │                                                                                                              
                                                                          no                                                                                                              
                                                                           ▼                                                                                                              
                                                            ┌────────────────────────────┐                                                                                                
                                                            │                            │                                                                                                
                                                      ┌─────│         I want to          │─────┐                                                                                          
                                                      │     │                            │     │                                                                                          
                                                      │     └────────────────────────────┘     │                                                                                          
                                                  Join an                                      │                                                                                          
                                                  existing                             Create a new team                                                                                  
                                                    team                                       │                                                                                          
                                                      │                                        │                                                                                          
┌────────────────────────────┐                        ▼                                        ▼                                                                                          
│ Nope, the only way to join │          ┌───────────────────────────┐           ┌────────────────────────────┐         ┌────────────────────────────┐         ┌──────────────────────────┐
│   an existing team is to   │          │                           │           │ Is the team name I try to  │         │    Am I member of this     │         │  Nope, you can reuse an  │
│     receive an invite      │◀─── no ──│     I have an invite      │           │use the same of an existing │── yes ─▶│       organization?        │── no ──▶│organization name only if │
│                            │          │                           │           │ (non-global) organization? │         │                            │         │you already belong to that│
└────────────────────────────┘          └───────────────────────────┘           └────────────────────────────┘         └────────────────────────────┘         └──────────────────────────┘
                                                      │                                        │                                      │                                                   
                                                    yes                                       no                                    yes                                                   
                                                      └────────────────────────────────────────┼──────────────────────────────────────┘                                                   
                                                                                               │                                                                                          
                                                                                               ▼                                                                                          
                                            ┌────────────────────────────┐       ┌───────────────────────────┐                                                                            
                                            │                            │       │      Is the season's      │                                                                            
                                            │ Nope, subscription status  │       │  subscription status to   │                                                                            
                                            │       should be open       │◀─ no ─│          "Open"?          │                                                                            
                                            │                            │       │                           │                                                                            
                                            └────────────────────────────┘       └───────────────────────────┘                                                                            
                                                                                               │                                                                                          
                                                                                             yes                                                                                          
                                                                                               ▼                                                                                          
                                                                                 ┌───────────────────────────┐                                                                            
                                                                                 │   If there are special    │                                                                            
                                                                                 │requirements (i.e., special│                                                                            
                                                                                 │email domain required), do │                                                                            
                                                                                 │       I meet them?        │                                                                            
                                                                                 └───────────────────────────┘                                                                            
                                                                                               │                                                                                          
                                                                                             yes                                                                                          
                                                                                               ▼                                                                                          
                                                                                         ┌───────────┐                                                                                    
                                                                                         │           │                                                                                    
                                                                                         │    Yes    │                                                                                    
                                                                                         │           │                                                                                    
                                                                                         └───────────┘                                                                                    
```

### PR DESCRIPTION
**Pull request by Z-a-r-a-k-i**
close #618

### CODE REVIEW

File: go/pkg/pwapi/api_team-send-invite.go [hunk 1]
```diff
@@ -55,23 +55,17 @@ func (svc *service) TeamSendInvite(ctx context.Context, in *TeamSendInvite_Input
...
-			UserID: invitedUserID,
-		}).
-		Where(&pwdb.Season{
-			ID: team.SeasonID,
-		}).
-		Joins("JOIN team ON team_member.team_id = team.id").
-		Joins("JOIN season ON team.season_id = season.id").
-		First(&existCheck).
+		Model(pwdb.TeamMember{}).
+		Joins("JOIN team on team.id = team_member.team_id AND team.season_id = ? AND team.deletion_status = ?", team.SeasonID, pwdb.DeletionStatus_Active).
```
Comment (moul) [inline go/pkg/pwapi/api_team-send-invite.go:62]:
```suggestion
		Joins("JOIN team on team.id = team_member.team_id").
		Where("team.season_id = ?", team.SeasonID).
		Where("team.deletion_status = ?", team.SeasonID, pwdb.DeletionStatus_Active).
```

Comment (moul) [reply go/pkg/pwapi/api_team-send-invite.go:62]: btw I've a doubt if it works to use the alternative `Where(Team{SeasonID: team.SeasonID})` format or if gorm will skip the field prefix

but anyway, please, split the where as I suggested

File: go/pkg/pwapi/api_team-send-invite.go [hunk 2]
```diff
@@ -55,23 +55,17 @@ func (svc *service) TeamSendInvite(ctx context.Context, in *TeamSendInvite_Input
...
+		Preload("Team").
```
Comment (moul) [inline go/pkg/pwapi/api_team-send-invite.go:63]:
```suggestion
		Preload("Team").
```

from what I can see, you don't use the variable, so preloading is useless and makes an unecessary request

btw in general you should consider using the `Count()` to just check that there are results

File: go/pkg/pwapi/api_team-create.go [hunk 3]
```diff
@@ -93,6 +96,20 @@ func (svc *service) TeamCreate(ctx context.Context, in *TeamCreate_Input) (*Team
 		organizationID = organization.ID
 	}
 
+	// check that user is member of the organization
```
Comment (moul) [inline go/pkg/pwapi/api_team-create.go:99]: that's only if an organization with the same name already existed before

in that case, yes you need to belong to the organization

in the other case, creating the team will also create a new organization in a lazy way

File: go/pkg/pwapi/api_team-accept-invite.go [hunk 4]
```diff
@@ -36,15 +36,38 @@ func (svc *service) TeamAcceptInvite(ctx context.Context, in *TeamAcceptInvite_I
...
+		OrganizationID: teamInvite.Team.OrganizationID,
+	}
+	err = svc.db.Transaction(func(tx *gorm.DB) error {
+		// create team member
+		err = tx.Create(&teamMember).Error
+		if err != nil {
+			return pwdb.GormToErrcode(err)
+		}
+		// create orga member
+		return tx.Create(&orgaMember).Error
```
Comment (moul) [inline go/pkg/pwapi/api_team-accept-invite.go:67]: can you make the `if err != nil` here too?

this way we will have the `pwdb.GormToErrcode` wrapping the error

File: go/pkg/pwapi/api_team-create.go (hunk 3 above)
Comment (Z-a-r-a-k-i) [reply go/pkg/pwapi/api_team-create.go:99]: If you check a few lines upper, the organization is automatically created if not existing, however if one organization with this slug/id exists then user needs to be a member, isn't it fine ?

Comment (moul) [reply go/pkg/pwapi/api_team-create.go:104]: ah yes cool, so:

1. put this in a transaction
2. add a `else` ot only check if the user is member of the organization if we didn't created a new org
3. yep, looks good for the rest, only check if user in the org with same slug :+1:

**Review by moul — APPROVED** [review]

**Pull request merged**